### PR TITLE
Parse GPX dive coordinates: Qt XML framework

### PR DESCRIPTION
**Pull request by willemferguson**
This replaces the C-code XML parsing with a Qt infrastructure.
QXmlStreamReader is used to parse the GPX file.

Signed-off-by: willemferguson <willemferguson@zoology.up.ac.za>

### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [ ] New feature
- [x] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Mentions:

@dirkhh @neolit123

### CODE REVIEW

File: core/parse-gpx.cpp [hunk 1]
```diff
@@ -86,61 +38,50 @@ int getCoordsFromGPXFile(struct dive_coords *coords, QString fileName)
...
-			first_line = false;
-			coords->start_track = when;   // Local time of start of GPS track
-		}
-		if ((when > divetime && found == false)) {   // This GPS local time corresponds to the start time of the dive
-			coords->lon = lon; // save the coordinates
-			coords->lat = lat;
-			found = true;
-		}
+	QXmlStreamReader gpxReader(&gpxFile);
+	while(!gpxReader.atEnd()) {
```
Comment (neolit123) [inline core/parse-gpx.cpp]: `while (`

File: core/parse-gpx.cpp [hunk 2]
```diff
@@ -86,61 +38,50 @@ int getCoordsFromGPXFile(struct dive_coords *coords, QString fileName)
...
+		gpxReader.readNext();
+		if (gpxReader.isStartElement()) {
+			if (gpxReader.name() == "trkpt") {
+				trkpt_found = true;
+				line++;
+				foreach(const QXmlStreamAttribute &attr, gpxReader.attributes()) {
```
Comment (neolit123) [inline core/parse-gpx.cpp]: `foreach (`

File: core/parse-gpx.cpp [hunk 3]
```diff
@@ -86,61 +38,50 @@ int getCoordsFromGPXFile(struct dive_coords *coords, QString fileName)
...
+					if (attr.name().toString() == QLatin1String("lat"))
+						lat = attr.value().toString().toDouble();
+					if (attr.name().toString() == QLatin1String("lon"))
```
Comment (neolit123) [inline core/parse-gpx.cpp]: `else if` ?

File: core/parse-gpx.cpp [hunk 4]
```diff
@@ -86,61 +38,50 @@ int getCoordsFromGPXFile(struct dive_coords *coords, QString fileName)
...
+						lon = attr.value().toString().toDouble();
+				}
+			}
+			if ((gpxReader.name() == "time") && trkpt_found){  // Ignore the <time> element in the GPX file header
```
Comment (neolit123) [inline core/parse-gpx.cpp]: `if (gpxReader.name() == "time" && trkpt_found) {`

File: core/parse-gpx.cpp [hunk 5]
```diff
@@ -86,61 +38,50 @@ int getCoordsFromGPXFile(struct dive_coords *coords, QString fileName)
...
+				tm1.tm_mday = dateTimeString.mid(8,2).toInt(&ok,10);
+				tm1.tm_hour = dateTimeString.mid(11,2).toInt(&ok,10);
+				tm1.tm_min  = dateTimeString.mid(14,2).toInt(&ok,10);
+				tm1.tm_sec  = dateTimeString.mid(17,2).toInt(&ok,10);
+				trkpt_time = utc_mktime(&tm1) + time_offset;
+				if (first_line) {
+					first_line = false;
+					coords->start_track = trkpt_time;   // Local time of start of GPS track
+				}
+				if ((trkpt_time >= divetime && found == false)) {   // This GPS local time corresponds to the start time of the dive
```
Comment (neolit123) [inline core/parse-gpx.cpp]: `if (trkpt_time >= divetime && found == false) {`

**Review by neolit123 — COMMENTED** [review]
thanks for the update @willemferguson 
this looks much simpler now.

added minor style comments,

**Comment by willemferguson** [comment]
Comments of @neolit123 have been responded to. Thank you, @neolit123.

**Comment by neolit123** [comment]
LGTM

**Pull request merged**